### PR TITLE
Add CI steps for IMEX build

### LINKED ISSUE #228: Add imex build to CI
Check the build and smoke tests for the linalg->xegpu->spirv lowering path.

### PR DESCRIPTION
**Pull request by dchigarev**
Closes #228 

Building graph compiler with IMEX requires LLVM with [applied patches](https://github.com/Menooker/mlir-extensions/tree/main/build_tools/patches).

This PR:
1. Adds `--imex` option for `scripts/build.sh` to build graph compiler with IMEX (and patched LLVM)
2. Adds `build-gpu-imex.yml` CI job, that builds GC with IMEX and runs tests
3. Adds `[CPU, IMEX]` matrix to `build-llvm.yml` CI job to build patched LLVM for IMEX

No unsupported tests mean, that all tests (including the ones for IMEX) have passed:
https://github.com/intel/graph-compiler/actions/runs/10390071602/job/28769671795?pr=245#step:7:59

p.s. we may eventually want to add a check that would verify that GPU tests are not marked as unsupported on GPU builds

### CODE REVIEW

**Comment by dchigarev** [comment]
@leshikus what's our policy to install dependencies for CI jobs? The IMEX/GPU build requires opencl to be installed on the runner and it [seems to be missing on our runners](https://github.com/intel/graph-compiler/actions/runs/10386239735/job/28757624962?pr=245#step:6:73). How do we typically handle those cases? Should I simply add a step in CI-job to install opencl before the build or should one of the runner's maintainers install opencl on the runner manually and only once?

**Comment by kurapov-peter** [comment]
blocked by #203

File: scripts/compile.sh
Comment (leshikus) [inline scripts/compile.sh:115]: ENABLE_IMEX is boolean; this is equivalent to if $ENABLE_IMEX; then

File: .github/workflows/build-llvm.yml
Comment (leshikus) [inline .github/workflows/build-llvm.yml:53]: this is likely debug, can be removed

**Review by leshikus — APPROVED** [review]
Have two minor comments, otherwise the patch is good

File: .github/workflows/build-llvm.yml
Comment (leshikus) [reply .github/workflows/build-llvm.yml:53]: the quotation is incorrect; debug is below this

**Review by Menooker — APPROVED** [review]

File: .github/workflows/build-llvm.yml
Comment (kurapov-peter) [inline .github/workflows/build-llvm.yml:66]: Is this enough to avoid artifact overlapping?

File: CMakeLists.txt
Comment (dchigarev) [reply CMakeLists.txt:56]: Non-normalized values cause problems in lit configs, where we determine if a variable is true or false [by comparing its value with "ON" or "1"](https://github.com/intel/graph-compiler/blob/88a282d923149144a59ea4208c7da73eef5166f2/test/mlir/test/lit.site.cfg.py.in#L37). It means, that specifying `-DGC_ENABLE_IMEX=On/True/TRUE/...` will be recognized as `TRUE` cmake, but as `FALSE` in the lit config.

In our lit configs we only use 2 bool variables: [`gc_use_imex`](https://github.com/intel/graph-compiler/blob/88a282d923149144a59ea4208c7da73eef5166f2/test/mlir/test/lit.site.cfg.py.in#L37) and [`gc_use_dnnl`](https://github.com/intel/graph-compiler/blob/88a282d923149144a59ea4208c7da73eef5166f2/test/mlir/test/lit.site.cfg.py.in#L46), meaning we should add normalization for both of them (will add normalization for `gc_use_dnnl` in the next commit)

File: .github/workflows/build-llvm.yml
Comment (dchigarev) [reply .github/workflows/build-llvm.yml:66]: I believe it should be enough. The artifacts will be stored under different names for imex and cpu builds, so they shouldn't overlap. @leshikus correct me if I'm wrong

File: .github/workflows/build-gpu-imex.yml
Comment (kurapov-peter) [reply .github/workflows/build-gpu-imex.yml:44]: Right, that's what I'd expect. Let's switch to the default packages if they are enough.

File: CMakeLists.txt
Comment (kurapov-peter) [reply CMakeLists.txt]: Ok, let's add some generic mechanism once we have 3 :)

**Review by kurapov-peter — APPROVED** [review]
LGTM, with the tweaks.

**Comment by dchigarev** [comment]
The LLVM build is now failing since the [new LLVM](https://github.com/intel/graph-compiler/commit/eeafe1a0c4e79799786864a0fef876c8a6754e75#diff-c7db5934bd7e72104442bdcdbbe2590cad06064694c2ac194e8a485ae5473449) is not compatible with IMEX, investigating

File: .github/workflows/build-gpu-imex.yml
Comment (leshikus) [reply .github/workflows/build-gpu-imex.yml:44]: I was stripped from admin rights, so I may fail to update the runner. Will try to replace a container only

Comment (leshikus) [reply .github/workflows/build-gpu-imex.yml:44]: The current status is: we no longer have GPU CI as for now. I was updating the runner and lost connect to the server

**Comment by dchigarev** [comment]
> The CI is now failing since the [new LLVM](https://github.com/intel/graph-compiler/commit/eeafe1a0c4e79799786864a0fef876c8a6754e75#diff-c7db5934bd7e72104442bdcdbbe2590cad06064694c2ac194e8a485ae5473449) is not compatible with IMEX, investigating

@kurapov-peter updating LLVM for IMEX this time means patching both [imex](https://github.com/Menooker/mlir-extensions/pull/4) and [linalg-to-xegpu pass in GC](https://github.com/dchigarev/graph-compiler/commit/e21f2421d70c06c6c01c9f9edc148d7192f129b4), so it would take sometime to be merged. That made me thinking of whether we should introduce two llvm pins (for CPU and IMEX builds) with a target to keep them identical.

This way we could make efforts to update LLVM in IMEX in parallel to patches for CPU mode (first update LLVM version for CPU and then later for IMEX). On the other side I understand that it makes things more complicated and also tempts us to "forget" updating LLVM for IMEX since CI will always be green even with old llvm.

I've added `llvm-version-imex.txt` in this PR to pin the last working LLVM version for imex. @kurapov-peter what do you think?

**Comment by kurapov-peter** [comment]
> > The CI is now failing since the [new LLVM](https://github.com/intel/graph-compiler/commit/eeafe1a0c4e79799786864a0fef876c8a6754e75#diff-c7db5934bd7e72104442bdcdbbe2590cad06064694c2ac194e8a485ae5473449) is not compatible with IMEX, investigating
> 
> @kurapov-peter updating LLVM for IMEX this time means patching both [imex](https://github.com/Menooker/mlir-extensions/pull/4) and [linalg-to-xegpu pass in GC](https://github.com/dchigarev/graph-compiler/commit/e21f2421d70c06c6c01c9f9edc148d7192f129b4), so it would take sometime to be merged. That made me thinking of whether we should introduce two llvm pins (for CPU and IMEX builds) with a target to keep them identical.
> 
> This way we could make efforts to update LLVM in IMEX in parallel to patches for CPU mode (first update LLVM version for CPU and then later for IMEX). On the other side I understand that it makes things more complicated and also tempts us to "forget" updating LLVM for IMEX since CI will always be green even with old llvm.
> 
> I've added `llvm-version-imex.txt` in this PR to pin the last working LLVM version for imex. @kurapov-peter what do you think?

Go for it. Having to work through the patching every time we need an llvm update is too much.

File: .github/workflows/build-llvm.yml
Comment (leshikus) [reply .github/workflows/build-llvm.yml:71]: @dchigarev correct

File: scripts/compile.sh
Comment (AndreyPavlenko) [inline scripts/compile.sh:34]: It will not work if -l is specified before -i.

File: scripts/compile.sh
Comment (dchigarev) [reply scripts/compile.sh:134]: works on my machine (c)

could you provide more details? from which folder do you launch the script? what's the path of `mlir-extensions` folder you end up with?

Comment (AndreyPavlenko) [reply scripts/compile.sh:134]: The problem is that llvm-project is a symbolic link. I'll create a PR with the fix.

**Pull request merged**
